### ISSUE TITLE
How to limit the number of CPU cores used? (Tensorflow backend)

### ISSUE THREAD

**Issue by maxmatical** [rstudio/keras#562]

**Comment by maxmatical** [comment]
I don't wish to disable parallelism completely. I am working with a server with 50 cpu cores, but since Keras takes up all of them, I wish to only use 10 or so. How can I do that?

**Comment by skeydan** [comment]
Does this work for you:

```
library(keras)
library(tensorflow)

config <- tf$ConfigProto(intra_op_parallelism_threads = 2L, inter_op_parallelism_threads = 2L)
session = tf$Session(config=config)
k_set_session(session)
```

**Comment by skeydan** [comment]
closing for inactivity; can always reopen if new information

**Comment by bastia1989** [comment]
> Does this work for you:
> 
> ```
> library(keras)
> library(tensorflow)
> 
> config <- tf$ConfigProto(intra_op_parallelism_threads = 2L, inter_op_parallelism_threads = 2L)
> session = tf$Session(config=config)
> k_set_session(session)
> ```

Would you explain why you set up this way. 10 cores are needed, and how this can be satisfied by defining `intra_op_parallelism_threads` and `inter_op_parallelism_threads`. Thanks.

**Comment by skeydan** [comment]
that was just an example... I meant the code per se :-)

**Comment by bastia1989** [comment]
> that was just an example... I meant the code per se :-)

So would you explain how exactly we can assign 10 CPU cores for it? I am not exactly sure here `intra_op_parallelism_threads` and `inter_op_parallelism_threads` are designed for CPU intensity. Thanks.

**Comment by skeydan** [comment]
There is an excellent explanation here: 

https://stackoverflow.com/a/41233901

In any case, I'd execute some tests to see how it works out for the kind of graphs you run, and your hardware.

**Comment by kmiddleton** [comment]
I'm trying to limit my core usage as well. Following the suggestion above, I have this code at the top of my .R file:

```
library(tensorflow)

config <- tf$ConfigProto(intra_op_parallelism_threads = 2L,
                         inter_op_parallelism_threads = 2L)
session = tf$Session(config = config)
k_set_session(session)
```

`fit()` is currently using all 56 cores on the machine. I think it should be using 2 (or maybe 4).

My session info is:

```
R version 3.5.1 (2018-07-02)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux

Matrix products: default
BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] tensorflow_1.10 cowplot_0.9.4   ggplot2_3.1.0   keras_2.2.4    

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.0       pillar_1.3.1     compiler_3.5.1   plyr_1.8.4       base64enc_0.1-3  tools_3.5.1     
 [7] digest_0.6.18    zeallot_0.1.0    jsonlite_1.6     tibble_2.0.1     gtable_0.2.0     lattice_0.20-38 
[13] pkgconfig_2.0.2  rlang_0.3.1      Matrix_1.2-15    rstudioapi_0.9.0 withr_2.1.2      dplyr_0.8.0.1   
[19] stringr_1.4.0    generics_0.0.2   grid_3.5.1       tidyselect_0.2.5 reticulate_1.10  glue_1.3.0      
[25] R6_2.4.0         purrr_0.3.0      reshape2_1.4.3   magrittr_1.5     whisker_0.3-2    scales_1.0.0    
[31] tfruns_1.4       assertthat_0.2.0 colorspace_1.4-0 labeling_0.3     stringi_1.3.1    lazyeval_0.2.1  
[37] munsell_0.5.0    crayon_1.3.4
```

**Comment by HaukurPall** [comment]
I am also unable to reduce the number of cores used by TensorFlow. These parameters seem to be ignored when I set them.

I am running r-keras inside a docker container on a host with many cores. TensorFlow tries to use all of them but is limited by docker (I want that) so it uses each core partly. This is causing serious performance issues.

I have tried clearing the session, set these config values and start a new session, but it does not seem to work.

Any ideas?
